### PR TITLE
[None][perf] Improve inference correctness and perf for Gemma4

### PR DESCRIPTION
**Pull request by 2ez4bz**
## Summary by CodeRabbit

* **New Features**
  * Added support for improved multimodal Gemma-4 handling, including image, video, and audio inputs.
  * Expanded B200 NVFP4 evaluation coverage for Gemma-4 multimodal runs.
* **Bug Fixes**
  * Improved attention and KV cache handling to reduce extra processing and better support paged cache indexing.
  * Refined tokenizer incremental decoding so it can recover from certain prefix-related decode errors.
  * Updated Gemma-4 routing and masking behavior for more consistent model outputs.

## Description

* Why?

Gemma 4 diverged from the Hugging Face reference for multimodal
attention and MoE routing, while chunked prefill repeated avoidable
encoder and mask work. Incremental decoding could also fail on invalid
byte prefixes, and padded KV page tables added unnecessary overhead.

* What?

This commit:

- aligns Gemma 4 masks and routing with Hugging Face, streamlines mask
  construction, and reuses multimodal embeddings across prefill chunks.
- recovers invalid HF decode streams and limit FlashInfer and KV cache
  page processing to each sequence's live block count.
- tweaks unit tests against the HF reference implementation to tighten
  bounds by an order of magnitude.

## Performance comparison

NVFP4, B200

### 26B

throughput_per_user_vs_total_output_throughput

### 31B

image

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Comment by nvpohanh** [comment]
[by Codex] @yizhang-nv Friendly reminder: could you please review this PR? Thanks!

File: tensorrt_llm/_torch/models/modeling_gemma4.py [hunk 1]
```diff
@@ -449,17 +449,14 @@ def apply(
...
     ) -> Tuple[torch.Tensor, torch.Tensor]:
-        # softmax over all experts
-        router_probs = F.softmax(router_logits.to(self.output_dtype), dim=-1)
-        # top-k selection
-        topk_weights, topk_indices = torch.topk(router_probs, k=self.top_k, dim=-1)
-        # renormalize so selected weights sum to 1
-        topk_weights = topk_weights / (topk_weights.sum(dim=-1, keepdim=True) + 1e-20)
-        # apply per-expert scale
+        # Preserve HF's operation order: BF16 softmax rounding can change top-k ties.
+        router_probabilities = F.softmax(router_logits, dim=-1)
```
Comment (Hudayday) [inline tensorrt_llm/_torch/models/modeling_gemma4.py]: Transformers uses float32 for all `softmax` operations (see [modular_gemma4.py#L1157](https://github.com/huggingface/transformers/blob/9f909e0f7027285d83bb6addc88155f1b80244ab/src/transformers/models/gemma4/modular_gemma4.py#L1157)). Do we need to align with them?

Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_gemma4.py]: Done.

File: tests/unittest/_torch/executor/test_per_layer_head_dim.py [hunk 2]
```diff
@@ -19,7 +19,11 @@
 
 import tensorrt_llm
 import tensorrt_llm.bindings
-from tensorrt_llm._torch.pyexecutor.kv_cache_manager_v2 import KVCacheManagerV2, Role
+from tensorrt_llm._torch.pyexecutor.kv_cache_manager_v2 import (
+    BAD_PAGE_INDEX,
+    KVCacheManagerV2,
+    Role,
+)
```
[the remark this answers is not in the document]
Comment (2ez4bz) [reply tests/unittest/_torch/executor/test_per_layer_head_dim.py:26]: Done.

File: tensorrt_llm/_torch/models/modeling_gemma4mm.py [hunk 3]
```diff
@@ -876,141 +878,157 @@ def _get_audio_features(
...
+            if self._has_active_multimodal_tokens(multimodal_param)
+        ]
+
+        mm_embeds = []
+        if active_multimodal_params:
+            mm_embeds = get_multimodal_embeddings(
+                encoder_forward_fn=self._forward_multimodal_encoder,
+                multimodal_params=active_multimodal_params,
+            )
+            mm_embeds = find_input_mm_embeds(mm_embeds, active_multimodal_params)
```
[the remark this answers is not in the document]
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_gemma4mm.py:1051]: Done.

**Review by lfr-0531 — APPROVED** [review]

**Review by Hudayday — APPROVED** [review]
LGTM.

**Review by yizhang-nv — APPROVED** [review]
LGTM on KVCM part

File: tensorrt_llm/_torch/models/modeling_gemma4mm.py [hunk 4]
```diff
@@ -876,141 +879,188 @@ def _get_audio_features(
...
-                if vsl is not None:
-                    video_seq_lens_extended.extend(vsl)
 
-        mm_embeds = []
-        all_mm_token_ids = []
-        mm_token_type_ids = None
-
-        # --- Process image features ---
-        if len(pixel_values_list) > 0:
+        for multimodal_param in multimodal_params:
```
Comment (yechank-nvidia) [inline tensorrt_llm/_torch/models/modeling_gemma4mm.py:947]: Not related to this PR, but I was wondering if we could this batching feature to be general enough that we can remove this if-else lines.

**Review by yechank-nvidia — APPROVED** [review]
LGTM

File: tensorrt_llm/_torch/models/modeling_gemma4.py [hunk 5]
```diff
@@ -1003,15 +1002,11 @@ def get_context_mask(
...
-        else:
-            pos = torch.arange(extend_len, device=device)
-            causal_mask = (pos.unsqueeze(0) <= pos.unsqueeze(1)) & (
-                pos.unsqueeze(0) > pos.unsqueeze(1) - effective_sliding_window
-            )
+        pos = torch.arange(extend_len, device=device)
+        causal_mask = pos.unsqueeze(0) <= pos.unsqueeze(1)
+
+        if effective_sliding_window is not None and effective_sliding_window < extend_len:
+            causal_mask.logical_and_(pos.unsqueeze(0) > pos.unsqueeze(1) - effective_sliding_window)
```
Comment (PerkzZheng) [inline tensorrt_llm/_torch/models/modeling_gemma4.py:1009]: curious that would this introduce two kernels in torch or this will be fused by torch compile ?

File: tensorrt_llm/_torch/models/modeling_gemma4mm.py (hunk 4 above)
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_gemma4mm.py:947]: Yep, noted. I would like this to be moved to the mixin eventually.

File: tensorrt_llm/_torch/models/modeling_gemma4mm.py [hunk 6]
```diff
@@ -876,141 +879,188 @@ def _get_audio_features(
...
+        """Whether a context parameter needs embeddings in this forward."""
+        runtime = multimodal_param.multimodal_runtime
+        if runtime is not None and runtime.num_mm_tokens_in_chunk == 0:
+            return False
+
+        multimodal_data = multimodal_param.multimodal_data
+        if multimodal_data.get("multimodal_embedding") is not None:
+            return True
+
+        payload_fields = (
```
Comment (eopXD) [inline tensorrt_llm/_torch/models/modeling_gemma4mm.py:893]: Do we need to handle un-expected field here to fail gracefully?

**Review by PerkzZheng — APPROVED** [review]

File: tensorrt_llm/_torch/models/modeling_gemma4.py (hunk 5 above)
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_gemma4.py:1009]: Good question. This mask construction is outside Gemma4’s current `torch.compile` scope, so I don't think there would be fusion here.

For the sliding-window case, the prior expression also performed two comparisons plus a logical-and. This rewrite keeps that work while using the existing causal-mask buffer in-place, avoiding an additional result allocation.

File: tensorrt_llm/_torch/models/modeling_gemma4mm.py (hunk 6 above)
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_gemma4mm.py:893]: I believe user-facing modalities are rejected earlier by `Gemma4InputProcessor._preprocess`, which validates `multi_modal_data` against image / audio / video and raises a `KeyError`.

At this layer, `multimodal_data` also carries valid metadata and cached embeddings, so `_has_active_multimodal_tokens` intentionally whitelists only the fields that require encoder work and ignores the rest.
